test/common/network/connection_impl_test.cc

```diff
@@ -583,23 +583,26 @@ TEST_P(ConnectionImplTest, IgnoreEmptyReadTest) {
   setUpBasicConnection();
   connect();
 
+  // Write some data and verify that it appears at the read filter. Leave the buffer undrained.
   const int buffer_size = 32;
   Buffer::OwnedImpl data(std::string(buffer_size, 'a'));
-  client_connection_->write(data);
-  client_connection_->close(ConnectionCloseType::FlushWrite);
-
   EXPECT_CALL(*read_filter_, onNewConnection());
   EXPECT_CALL(*read_filter_, onData(_))
       .Times(1)
       .WillOnce(Invoke([&](Buffer::Instance& data) -> FilterStatus {
         EXPECT_EQ(buffer_size, data.length());
+        dispatcher_->exit();
         return FilterStatus::StopIteration;
       }));
+  client_connection_->write(data);
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
 
+  // Close the connection and verify that the undrained buffer is not passed to the read filter a
+  // second time.
   EXPECT_CALL(client_callbacks_, onEvent(ConnectionEvent::LocalClose));
   EXPECT_CALL(server_callbacks_, onEvent(ConnectionEvent::RemoteClose))
       .WillOnce(Invoke([&](Network::ConnectionEvent) -> void { dispatcher_->exit(); }));
-
+  client_connection_->close(ConnectionCloseType::NoFlush);
```

Comment:
Oh sorry for miscommunication but I think we want both tests, the original and this edited one.  It'd be good to verify if the data and fin arrive together (which they usually do on non-mac-OS) we handle that case as well as this.  WDYT?

Reply:
Ah. Ok.

Reply:
Comments wrong now anyway, since in this version of the test there's always an empty read, even on Linux.